dsl/src/main/scala/com.crobox.clickhouse/dsl/AggregationFunctions.scala

```diff
@@ -0,0 +1,259 @@
+  case class State[T <: Column]()                             extends Combinator[T, StateResult]
+  case class Merge[T <: TableColumn[StateResult], Res]()      extends Combinator[T, Res]
+
+  trait AggregationFunctionsCombinersDsl {
+
+    def aggIf[T <: TableColumn[Res], Res](condition: Comparison)(aggregated: AggregateFunction[Res]) =
+      CombinedAggregatedFunction(If(condition), aggregated)
+
+    def array[T <: TableColumn[Seq[Res]], Res](aggregated: AggregateFunction[Res]) =
+      CombinedAggregatedFunction(CombinatorArray[T, Res](), aggregated)
+
+    def forEach[T <: TableColumn[Seq[Seq[V]]], V](aggregated: AggregateFunction[V]) =
+      CombinedAggregatedFunction(ArrayForEach[T, V](), aggregated)
+
+    def state[T <: TableColumn[_]](aggregated: AggregateFunction[_]) =
+      CombinedAggregatedFunction(State[T](), aggregated)
+
+    def merge[T <: TableColumn[StateResult], Res](aggregated: AggregateFunction[Res]) =
+      CombinedAggregatedFunction(Merge[T, Res](), aggregated)
+  }
+
+  trait AggregationFunctionsDsl
+      extends AggregationFunctionsCombinersDsl
+      with UniqDsl
+      with AnyResultDsl
+      with SumDsl
+      with LevelModifierDsl {
+
+    def count() =
+      Count()
+
+    def count(column: TableColumn[_]) =
+      Count(Option(column))
+
+    def uniq(tableColumn: AnyTableColumn) =
+      Uniq(tableColumn)
+
+    def average[T: Numeric](tableColumn: TableColumn[T]) =
+      Avg(tableColumn)
+
+    def any[T](tableColumn: TableColumn[T]) =
+      AnyResult(tableColumn)
+
+    def sum[T: Numeric](tableColumn: TableColumn[T]) =
+      Sum(tableColumn)
+
+    def min[V](tableColumn: TableColumn[V]) =
+      Min(tableColumn)
+
+    def max[V](tableColumn: TableColumn[V]) =
+      Max(tableColumn)
+
+    def timeSeries(tableColumn: TableColumn[Long],
+                   interval: MultiInterval,
+                   dateColumn: Option[TableColumn[DateTime]] = None) =
+      TimeSeries(tableColumn, interval, dateColumn)
+
+    def groupUniqArray[V](tableColumn: TableColumn[V]) =
+      GroupUniqArray(tableColumn)
+
+    def median[V](target: TableColumn[V], level: Float = 0.5F) = Median(target, level = level)
+
+    def quantile[V](target: TableColumn[V], level: Float = 0.5F) = Quantile(target, level = level)
+
+    def quantiles[V](target: TableColumn[V], levels: Float*) = Quantiles(target, levels)
+
+  }
+}
+
+case class Count(column: Option[AnyTableColumn] = None) extends AggregateFunction[Long](column.getOrElse(EmptyColumn()))
+
+case class Uniq(tableColumn: AnyTableColumn, modifier: UniqModifier = Uniq.Normal)
+    extends AggregateFunction[Long](tableColumn)
+
+object Uniq {
+  sealed trait UniqModifier
+  case object Normal   extends UniqModifier
+  case object Combined extends UniqModifier
+  case object HLL12    extends UniqModifier
+  case object Exact    extends UniqModifier
+
+  trait UniqDsl {
+    def combined(uniq: Uniq) = uniq.copy(modifier = Combined)
```

Comment:
We could provide all the possible combinations as clickhouse for uniq/sum/etc.. but having them apply on top as modifiers makes it more flexible for the library user, as it can create his own composed state of the aggregation and use that simply everywhere